README.md
# minutes.js

[![Travis](https://img.shields.io/travis/smebberson/minutes.js.svg)](smebberson/minutes.js)
[![npm](https://img.shields.io/npm/v/minutes.js.svg)](smebberson/minutes.js)
[![codecov.io](https://codecov.io/github/smebberson/minutes.js/coverage.svg?branch=master)](https://codecov.io/github/smebberson/minutes.js?branch=master)

Minutes.js is a JavaScript library for working with minutes. Put minutes in, and get a string representing those minutes in weeks, days, hours and minutes.

## Install

```
$ npm install minutes.js
```

Supports browser, Node.js and AMD (and ES6 packages via `src/minutes.js`).

## API

```
minutes(10);                        // 10 minutes
minutes(2*60+50);                   // 2 hours and 50 minutes
minutes(2*24*60+9*60+30);           // 2 days, 6 hours and 30 minutes
minutes(2*7*24*60+6*24*60+7*60+40); // 2 weeks, 6 days, 7 hours and 40 minutes
```

### Minutes

Minutes takes a `minutes` number. It should be an integer representing minutes.

### Options

Minutes takes an `options` Object. They're all optional.

#### Units

`units` is an `Object` that will define the strings used to represent the units. By default, these strings are verbose:

```
units: {
    m: 'minute',
    h: 'hour',
    d: 'day',
    w: 'week'
}
```

You can customise these to anything you like:

```
const opts = {
    units: {
        m: 'min',
        h: 'hr',
        d: 'day',
        w: 'wk'
    }
};

minutes(10, opts);                          // 10 mins
minutes(2*60+50, opts);                     // 2 hrs and 50 mins
minutes(2*24*60+9*60+30, opts);             // 2 days, 6 hrs and 30 mins
minutes(2*7*24*60+6*24*60+7*60+40, opts);   // 2 wks, 6 days, 7 hrs and 40 mins
```

#### Pluralize

`pluralize` is a `boolean` that will determine if the strings used to represent a unit should be pluralized if there is more than one of a particular unit. `pluralize` is `true` by default to produce `2 hours and 1 minute`.

You can alter this:

```
const opts = {
    units: {
        m: 'min',
        h: 'hr',
        d: 'day',
        w: 'wk'
    },
    pluralize: false
};

minutes(10, opts);                          // 10 min
minutes(2*60+50, opts);                     // 2 hr and 50 min
minutes(2*24*60+9*60+30, opts);             // 2 day, 6 hr and 30 min
minutes(2*7*24*60+6*24*60+7*60+40, opts);   // 2 wk, 6 day, 7 hr and 40 min
```

#### Tokens

`tokens` is an `Object` that represents the strings used between the various units. The default is:

```
tokens: {
    conjunction: 'and ',
    delimiter: ', ',
    plural: 's',
    space: ' '
}
```

You can customise this:

```
const opts = {
    units: {
        m: 'm',
        h: 'h',
        d: 'd',
        w: 'w'
    },
    pluralize: false,
    tokens: {
        conjunction: ' ',
        delimiter: ' ',
        space: ''
    }
};

minutes(10, opts);                          // 10m
minutes(2*60+50, opts);                     // 2h 50m
minutes(2*24*60+9*60+30, opts);             // 2d 6h 30m
minutes(2*7*24*60+6*24*60+7*60+40, opts);   // 2w 6d 7h 40m
```

Or something else:

```
const opts = {
    units: {
        m: '',
        h: '',
        d: '',
        w: ''
    },
    pluralize: false,
    tokens: {
        conjunction: '.',
        delimiter: '.',
        space: ''
    }
};

minutes(10, opts);                          // 10m
minutes(2*60+50, opts);                     // 2h:50m
minutes(2*24*60+9*60+30, opts);             // 2d:6h:30m
minutes(2*7*24*60+6*24*60+7*60+40, opts);   // 2w:6h:7h:40m
```

#### Display

`display` is an `Object` which dictates the units to be displayed. It defaults to all units:

```
display: {
    m: true,
    h: true,
    d: true,
    w: true
}
```

You can change this to display only certain units:

```

minutes(10, {
    display: {
        h: false,
        d: false,
        w: false
    }
});                                     // 10 minutes
minutes(2*60+50, {
    display: {
        m: false,
        d: false,
        w: false
    }
});                                     // 2 hours
minutes(1*24*60+9*60+30, {
    display: {
        m: false,
        h: false,
        w: false
    }
});                                     // 1 day
minutes(3*7*24*60+6*24*60+7*60+40, {
    display: {
        m: false,
        h: false,
        d: false
    }
});                                     // 3 weeks
```

## Change log

[Review the change log for all changes.](CHANGELOG.md)

## License

[MIT](LICENSE.md)


CHANGELOG.md
# minutes.js changelog

All notable changes to minutes.js will be documented here. This project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased

-   Upgraded Rollup and Babel.
-   Added Prettier.
-   Updated code to use more es6 functionality (via babel).
-   Updated test code to use more es6 functionality (via babel).
-   Removed mocha and chai, replacing them with Jest.
-   Removed files that are no longer required.
-   Removed usage of Class, minutes is now a single function.
-   `options.tokens.comma` is now `options.tokens.delimiter`.
-   `options.tokens.and` is now `options.tokens.conjunction`.
-   Added double-digit display capabilities to prefix units with leading zeros.
-   Added inclusive display capabilities for units to display even when 0.
-   Added preset options to automate common options configurations.

### Breaking changes

-   The export has been renamed from `Minutes` to `minutes`.
-   You can no longer use `new` with minutes, it is now a single function.
-   The `toString()` method has been removed, instead, use the single function.
-   `options.tokens.comma` is now `options.tokens.delimiter`.
-   `options.tokens.and` is now `options.tokens.conjunction`.

## [1.1.1] - 2016-05-27

-   Falling back to `isNaN` if `Number.isNaN` is not present (to support IE).

## [1.1.0] - 2016-05-05

-   Added the `display` option to control which units are output.

## [1.0.1] - 2016-04-23

-   Now provides support for browser, Node.js and AMD.

## [1.0.0] - 2016-04-23

-   First complete version.


test/minutes.js
'use strict';

let expect = require('chai').expect,
    clone = require('clone'),
    Minutes = require('../dist/minutes');

describe('Minutes', () => {

    //
    // Constants
    //

    const HOUR = 60;

    it('will throw if not passed an integer', function () {

        var fn = function () {
            new Minutes('asdf');
        }

        expect(fn).to.throw(Error, /accepts an integer/);

    });

    it('will throw if executed without new', function () {

        var fn = function () {
            Minutes(23); // eslint-disable-line new-cap
        }

        expect(fn).to.throw(Error, /cannot/i);

    });

    describe('will format', function () {

        describe('single time units', function () {

            var minute = new Minutes(1),
                minutes = new Minutes(10),
                hour = new Minutes(1*60),
                hours = new Minutes(2*60),
                day = new Minutes(24*60),
                days = new Minutes(2*24*60),
                week = new Minutes(7*24*60),
                weeks = new Minutes(2*7*24*60);

            it('a minute', function () {

                expect(minute.toString()).to.equal('1 minute');

            });

            it('minutes', function () {

                expect(minutes.toString()).to.equal('10 minutes');

            });

            it('an hour', function () {

                expect(hour.toString()).to.equal('1 hour');

            });

            it('hours', function () {

                expect(hours.toString()).to.equal('2 hours');

            });

            it('a day', function () {

                expect(day.toString()).to.equal('1 day');

            });

            it('days', function () {

                expect(days.toString()).to.equal('2 days');

            });

            it('a week', function () {

                expect(week.toString()).to.equal('1 week');

            });

            it('weeks', function () {

                expect(weeks.toString()).to.equal('2 weeks');

            });

        });

        describe('multiple time units', function () {

           it('hours and minutes', function () {

                var hours = new Minutes(2*HOUR+10);

                expect(hours.toString()).to.equal('2 hours and 10 minutes');

            });

            it('days, hours and minutes', function () {

                var hours = new Minutes((3*24*60)+(2*60)+10);

                expect(hours.toString()).to.equal('3 days, 2 hours and 10 minutes');

            });

            it('weeks, days, hours and minutes', function () {

                var hours = new Minutes((2*7*24*60)+(3*24*60)+(2*60)+10);

                expect(hours.toString()).to.equal('2 weeks, 3 days, 2 hours and 10 minutes');

            });

        });

        describe('using custom time unit strings', function () {

            describe('with plurals (default)', function () {

                var opts = {
                        units: {
                            'm': 'min',
                            'h': 'hr',
                            'd': 'day',
                            'w': 'wk'
                        }
                    },
                    minutes = new Minutes(10, opts),
                    hours = new Minutes(2*60+50, opts),
                    days = new Minutes(2*24*60+9*60+30, opts),
                    weeks = new Minutes(2*7*24*60+6*24*60+7*60+40, opts);

                it('for minutes', function () {

                    expect(minutes.toString()).to.equal('10 mins');

                });

                it('for hours', function () {

                    expect(hours.toString()).to.equal('2 hrs and 50 mins');

                });

                it('for days', function () {

                    expect(days.toString()).to.equal('2 days, 9 hrs and 30 mins');

                });

                it('for weeks', function () {

                    expect(weeks.toString()).to.equal('2 wks, 6 days, 7 hrs and 40 mins');

                });

            });

            describe('without plurals', function () {

                var opts = {
                        units: {
                            'm': 'm',
                            'h': 'h',
                            'd': 'd',
                            'w': 'w'
                        },
                        pluralize: false
                    },
                    minutes = new Minutes(10, opts),
                    hours = new Minutes(2*60+50, opts),
                    days = new Minutes(2*24*60+9*60+30, opts),
                    weeks = new Minutes(2*7*24*60+6*24*60+7*60+40, opts);

                it('for minutes', function () {

                    expect(minutes.toString()).to.equal('10 m');

                });

                it('for hours', function () {

                    expect(hours.toString()).to.equal('2 h and 50 m');

                });

                it('for days', function () {

                    expect(days.toString()).to.equal('2 d, 9 h and 30 m');

                });

                it('for weeks', function () {

                    expect(weeks.toString()).to.equal('2 w, 6 d, 7 h and 40 m');

                });

            });

        });

        describe('using custom tokens', function () {

            var opts = {
                    units: {
                        'm': 'm',
                        'h': 'h',
                        'd': 'd',
                        'w': 'w'
                    },
                    pluralize: false,
                    tokens: {
                        space: '',
                        comma: ' ',
                        and: ' '
                    }
                },
                minutes = new Minutes(9, opts),
                hours = new Minutes(5*60+23, opts),
                days = new Minutes(6*24*60+3*60+12, opts),
                weeks = new Minutes(6*7*24*60+3*24*60+2*60+55, opts);

            it('for minutes', function () {

                expect(minutes.toString()).to.equal('9m');

            });

            it('for hours', function () {

                expect(hours.toString()).to.equal('5h 23m');

            });

            it('for days', function () {

                expect(days.toString()).to.equal('6d 3h 12m');

            });

            it('for weeks', function () {

                expect(weeks.toString()).to.equal('6w 3d 2h 55m');

            });

            it('with regex sensitive characters', function () {

                opts = {
                    units: {
                        'm': 'm',
                        'h': 'h',
                        'd': 'd',
                        'w': 'w'
                    },
                    pluralize: false,
                    tokens: {
                        space: '',
                        comma: '^(?:=\\.*).+|$&$`$',
                        and: '^(?:=\\.*).+|$&$`$'
                    }
                };
                minutes = new Minutes(9, opts);
                hours = new Minutes(5*60+23, opts);
                days = new Minutes(6*24*60+3*60+12, opts);
                weeks = new Minutes(6*7*24*60+3*24*60+2*60+55, opts);

                expect(weeks.toString()).to.equal('6w^(?:=\\.*).+|$&$`$3d^(?:=\\.*).+|$&$`$2h^(?:=\\.*).+|$&$`$55m');

            });

        });

        describe('using custom display options', function () {

            var opts = {
                    units: {
                        'm': 'm',
                        'h': 'h',
                        'd': 'd',
                        'w': 'w'
                    },
                    pluralize: false,
                    tokens: {
                        space: '',
                        comma: ' ',
                        and: ' '
                    }
                },
                timespan = 6*7*24*60+3*24*60+2*60+55,
                minutes = new Minutes(timespan, Object.assign(clone(opts), {
                    display: {
                        'h': false,
                        'd': false,
                        'w': false
                    }
                })),
                hours = new Minutes(timespan, Object.assign(clone(opts), {
                    display: {
                        'd': false,
                        'w': false
                    }
                })),
                hoursOnly = new Minutes(timespan, Object.assign(clone(opts), {
                    display: {
                        'm': false,
                        'd': false,
                        'w': false
                    }
                })),
                days = new Minutes(timespan, Object.assign(clone(opts), {
                    display: {
                        'w': false
                    }
                })),
                daysOnly = new Minutes(timespan, Object.assign(clone(opts), {
                    display: {
                        'm': false,
                        'h': false,
                        'w': false
                    }
                })),
                weeks = new Minutes(timespan, opts),
                weeksOnly = new Minutes(timespan, Object.assign(clone(opts), {
                    display: {
                        'm': false,
                        'h': false,
                        'd': false
                    }
                }));

            it('minutes only', function () {

                expect(minutes.toString()).to.equal('64975m');

            });

            it('hours only', function () {

                expect(hoursOnly.toString()).to.equal('1082h');

            });

            it('days only', function () {

                expect(daysOnly.toString()).to.equal('45d');

            });

            it('weeks only', function () {

                expect(weeksOnly.toString()).to.equal('6w');

            });

            it('hours and minutes only', function () {

                expect(hours.toString()).to.equal('1082h 55m');

            });

            it('days, hours and minutes only', function () {

                expect(days.toString()).to.equal('45d 2h 55m');

            });

            it('weeks, days, hours and minutes', function () {

                expect(weeks.toString()).to.equal('6w 3d 2h 55m');

            });

        });

    });

});


test/minutes.test.js
'use strict';

import { default as minutes, HOUR } from '../src/minutes';

describe('minutes', () => {
    test('will throw if not passed an integer', () => {
        const fn = () => {
            minutes('asdf');
        };

        expect(fn).toThrow(Error, /accepts an integer/);
    });
    describe('will use Number.isNan', () => {
        const originalNumberIsNan = Number.isNaN;

        beforeAll(() => {
            Number.isNaN = undefined;
        });

        afterAll(() => {
            Number.isNaN = originalNumberIsNan;
        });
        test('before defaulting to isNaN', () => {
            isNaN = Number.isNaN;
            Number.isNaN = undefined;

            const fn = () => {
                minutes('asdf');
            };

            expect(fn).toThrow(Error, /accepts an integer/);

            Number.isNaN = isNaN;
        });
    });
});

describe('minutes will format single time units', () => {
    test('a minute', () => {
        expect(minutes(1)).toEqual('1 minute');
    });

    test('minutes', () => {
        expect(minutes(10)).toEqual('10 minutes');
    });

    test('an hour', () => {
        expect(minutes(60)).toEqual('1 hour');
    });

    test('hours', () => {
        expect(minutes(2 * 60)).toEqual('2 hours');
    });

    test('a day', () => {
        expect(minutes(24 * 60)).toEqual('1 day');
    });

    test('days', () => {
        expect(minutes(2 * 24 * 60)).toEqual('2 days');
    });

    test('a week', () => {
        expect(minutes(7 * 24 * 60)).toEqual('1 week');
    });

    test('weeks', () => {
        expect(minutes(2 * 7 * 24 * 60)).toEqual('2 weeks');
    });
});

describe('minutes will format multiple time units', () => {
    test('hours and minutes', () => {
        expect(minutes(2 * HOUR + 10)).toEqual('2 hours and 10 minutes');
    });

    test('days, hours and minutes', () => {
        expect(minutes(3 * 24 * 60 + 2 * 60 + 10)).toEqual(
            '3 days, 2 hours and 10 minutes'
        );
    });

    test('weeks, days, hours and minutes', () => {
        expect(minutes(2 * 7 * 24 * 60 + 3 * 24 * 60 + 2 * 60 + 10)).toEqual(
            '2 weeks, 3 days, 2 hours and 10 minutes'
        );
    });
});

describe('minutes will format using custom time unit strings', () => {
    describe('with plurals (default)', () => {
        const opts = {
            units: {
                m: 'min',
                h: 'hr',
                d: 'day',
                w: 'wk'
            }
        };

        test('for minutes', () => {
            expect(minutes(10, opts)).toEqual('10 mins');
        });

        test('for hours', () => {
            expect(minutes(2 * 60 + 50, opts)).toEqual('2 hrs and 50 mins');
        });

        test('for hours and no minutes', () => {
            expect(
                minutes(2 * 60, {
                    ...opts,
                    display: { inclusive: true, d: false, w: false }
                })
            ).toEqual('2 hrs and 0 mins');
        });

        test('for days', () => {
            expect(minutes(2 * 24 * 60 + 9 * 60 + 30, opts)).toEqual(
                '2 days, 9 hrs and 30 mins'
            );
        });

        test('for weeks', () => {
            expect(
                minutes(2 * 7 * 24 * 60 + 6 * 24 * 60 + 7 * 60 + 40, opts)
            ).toEqual('2 wks, 6 days, 7 hrs and 40 mins');
        });
    });

    describe('without plurals', () => {
        const opts = {
            units: {
                m: 'm',
                h: 'h',
                d: 'd',
                w: 'w'
            },
            pluralize: false
        };

        test('for minutes', () => {
            expect(minutes(10, opts)).toEqual('10 m');
        });

        test('for hours', () => {
            expect(minutes(2 * 60 + 50, opts)).toEqual('2 h and 50 m');
        });

        test('for days', () => {
            expect(minutes(2 * 24 * 60 + 9 * 60 + 30, opts)).toEqual(
                '2 d, 9 h and 30 m'
            );
        });

        test('for weeks', () => {
            expect(
                minutes(2 * 7 * 24 * 60 + 6 * 24 * 60 + 7 * 60 + 40, opts)
            ).toEqual('2 w, 6 d, 7 h and 40 m');
        });
    });
});

describe('minutes will format using custom tokens', () => {
    const opts = {
        units: {
            m: 'm',
            h: 'h',
            d: 'd',
            w: 'w'
        },
        pluralize: false,
        tokens: {
            conjunction: ' ',
            delimiter: ' ',
            space: ''
        }
    };

    test('for minutes', () => {
        expect(minutes(9, opts)).toEqual('9m');
    });

    test('for hours', () => {
        expect(minutes(5 * 60 + 23, opts)).toEqual('5h 23m');
    });

    test('for days', () => {
        expect(minutes(6 * 24 * 60 + 3 * 60 + 12, opts)).toEqual('6d 3h 12m');
    });

    test('for weeks', () => {
        expect(
            minutes(6 * 7 * 24 * 60 + 3 * 24 * 60 + 2 * 60 + 55, opts)
        ).toEqual('6w 3d 2h 55m');
    });

    test('with regex sensitive characters', () => {
        const opts = {
            units: {
                m: 'm',
                h: 'h',
                d: 'd',
                w: 'w'
            },
            pluralize: false,
            tokens: {
                conjunction: '^(?:=\\.*).+|$&$`$',
                delimiter: '^(?:=\\.*).+|$&$`$',
                space: ''
            }
        };

        expect(
            minutes(6 * 7 * 24 * 60 + 3 * 24 * 60 + 2 * 60 + 55, opts)
        ).toEqual(
            '6w^(?:=\\.*).+|$&$`$3d^(?:=\\.*).+|$&$`$2h^(?:=\\.*).+|$&$`$55m'
        );
    });
});

describe('minutes will format using inclusive display options', () => {
    test('inclusive hours', () => {
        expect(
            minutes(50, {
                display: {
                    inclusive: true,
                    d: false,
                    w: false
                },
                pluralize: false,
                tokens: {
                    conjunction: ':',
                    delimiter: '',
                    space: ''
                },
                units: {
                    m: '',
                    h: '',
                    d: '',
                    w: ''
                }
            })
        ).toEqual('0:50');
    });
    test('inclusive days and hours', () => {
        expect(
            minutes(2 * 60 + 50, {
                display: {
                    inclusive: true,
                    w: false
                },
                pluralize: true,
                tokens: {
                    conjunction: ' ',
                    delimiter: ' ',
                    space: ''
                },
                units: {
                    m: 'min',
                    h: 'hr',
                    d: 'day'
                }
            })
        ).toEqual('0days 2hrs 50mins');
    });
    test('inclusive weeks, days and hours', () => {
        expect(
            minutes(2 * 60 + 50, {
                display: {
                    inclusive: true
                },
                pluralize: true,
                tokens: {
                    conjunction: ' ',
                    delimiter: ' ',
                    space: ''
                },
                units: {
                    m: 'min',
                    h: 'hr',
                    d: 'day',
                    w: 'wk'
                }
            })
        ).toEqual('0wks 0days 2hrs 50mins');
    });
});

describe('minutes will format using double-digit display options', () => {
    test('double digit minutes', () => {
        expect(minutes(9, { display: { mm: true } })).toEqual('09 minutes');
    });

    test('double digit hours and minutes', () => {
        expect(minutes(129, { display: { hh: true, mm: true } })).toEqual(
            '02 hours and 09 minutes'
        );
    });

    test('double digit days, hours and minutes', () => {
        expect(
            minutes(2 * 24 * 60 + 129, {
                display: { dd: true, hh: true, mm: true }
            })
        ).toEqual('02 days, 02 hours and 09 minutes');
    });

    test('double digit weeks, days, hours and minutes', () => {
        expect(
            minutes(4 * 7 * 24 * 60 + 6 * 24 * 60 + 7 * 60 + 7, {
                display: { dd: true, hh: true, mm: true, ww: true }
            })
        ).toEqual('04 weeks, 06 days, 07 hours and 07 minutes');
    });
});

describe('minutes will format using custom display options', () => {
    const opts = {
        units: {
            m: 'm',
            h: 'h',
            d: 'd',
            w: 'w'
        },
        pluralize: false,
        tokens: {
            conjunction: ' ',
            delimiter: ' ',
            space: ''
        }
    };
    const timespan = 6 * 7 * 24 * 60 + 3 * 24 * 60 + 2 * 60 + 55;

    test('minutes only', () => {
        expect(
            minutes(timespan, {
                ...opts,
                display: {
                    h: false,
                    d: false,
                    w: false
                }
            })
        ).toEqual('64975m');
    });

    test('hours only', () => {
        expect(
            minutes(timespan, {
                ...opts,
                display: {
                    m: false,
                    d: false,
                    w: false
                }
            })
        ).toEqual('1082h');
    });

    test('days only', () => {
        expect(
            minutes(timespan, {
                ...opts,
                display: {
                    m: false,
                    h: false,
                    w: false
                }
            })
        ).toEqual('45d');
    });

    test('weeks only', () => {
        expect(
            minutes(timespan, {
                ...opts,
                display: {
                    m: false,
                    h: false,
                    d: false
                }
            })
        ).toEqual('6w');
    });

    test('hours and minutes only', () => {
        expect(
            minutes(timespan, {
                ...opts,
                display: {
                    d: false,
                    w: false
                }
            })
        ).toEqual('1082h 55m');
    });

    test('days, hours and minutes only', () => {
        expect(
            minutes(timespan, {
                ...opts,
                display: {
                    w: false
                }
            })
        ).toEqual('45d 2h 55m');
    });

    test('weeks, days, hours and minutes', () => {
        expect(minutes(timespan, opts)).toEqual('6w 3d 2h 55m');
    });

    test('expressed as time', () => {
        expect(
            minutes(timespan, {
                ...opts,
                display: {
                    d: false,
                    w: false
                },
                pluralize: false,
                tokens: {
                    conjunction: '',
                    delimiter: '',
                    space: ''
                },
                units: {
                    m: '',
                    h: ':',
                    d: '',
                    w: ''
                }
            })
        ).toEqual('1082:55');
        expect(
            minutes(timespan + 2, {
                ...opts,
                display: {
                    d: false,
                    w: false
                },
                pluralize: false,
                tokens: {
                    conjunction: ':',
                    delimiter: '',
                    space: ''
                },
                units: {
                    m: '',
                    h: '',
                    d: '',
                    w: ''
                }
            })
        ).toEqual('1082:57');
    });
});

describe('minutes will automate options using presets', () => {
    test('the time preset', () => {
        expect(minutes(9, { preset: 'time' })).toEqual('00:09');
    });
    test('the time preset with overrides', () => {
        expect(
            minutes(9, { preset: 'time', units: { h: 'h', m: 'm' } })
        ).toEqual('00h:09m');
    });
});


src/minutes.js
//
// Constants
//

export const HOUR = 60;
export const DAY = 24 * HOUR;
export const WEEK = 7 * DAY;

const presets = {
    time: {
        display: {
            inclusive: true,
            d: false,
            hh: true,
            mm: true,
            w: false
        },
        pluralize: false,
        tokens: {
            conjunction: ':',
            delimiter: '',
            space: ''
        },
        units: {
            m: '',
            h: '',
            d: '',
            w: ''
        }
    }
};

const applyPresets = (opts) => {
    if (opts.preset && Object.keys(presets).includes(opts.preset)) {
        return { ...presets[opts.preset], ...opts };
    }

    return opts;
};

const defaultOptions = (defaults) => {
    const opts = applyPresets({ ...defaults });

    // Default the time units.
    opts.units = opts.units ?? {};
    opts.units.w = opts.units.w ?? 'week';
    opts.units.d = opts.units.d ?? 'day';
    opts.units.h = opts.units.h ?? 'hour';
    opts.units.m = opts.units.m ?? 'minute';

    // Default the time units to display.
    opts.display = opts.display ?? {};
    opts.display.inclusive = opts.display.inclusive ?? false;
    opts.display.w = opts.display.w ?? true;
    opts.display.ww = opts.display.ww ?? false;
    opts.display.d = opts.display.d ?? true;
    opts.display.dd = opts.display.dd ?? false;
    opts.display.h = opts.display.h ?? true;
    opts.display.hh = opts.display.hh ?? false;
    opts.display.m = opts.display.m ?? true;
    opts.display.mm = opts.display.mm ?? false;

    // Default pluralize (to true).
    opts.pluralize = opts.pluralize ?? true;

    // Default separation tokens.
    opts.tokens = opts.tokens ?? {};
    opts.tokens.space = opts.tokens.space ?? ' ';
    opts.tokens.delimiter = opts.tokens.delimiter ?? ', ';
    opts.tokens.conjunction = opts.tokens.conjunction ?? ' and ';
    opts.tokens.plural = opts.tokens.plural ?? 's';

    return opts;
};

const padWithZero = (value, leadingZero) =>
    leadingZero && value < 10 ? `0${value}` : value;

/**
 * Format a time unit (i.e. 1 h as '1 hour').
 * @param  {Number} value The value of the time unit.
 * @param  {String} unit  The time unit (either `'w'`, `'h'`, `'m'`, `'d'`).
 * @return {String}       The string representation of the time unit (i.e. 5 hours).
 */
const formatPart = (value, unit, leadingZero, opts) => {
    let str =
        padWithZero(value, leadingZero) + opts.tokens.space + opts.units[unit];

    // Make the unit representation plural if required.
    if (opts.pluralize && (value > 1 || value === 0)) {
        str += opts.tokens.plural;
    }

    return str;
};

/**
 * Using the options provided to the constructor take the value of the minutes and
 * format into a string.
 * @return {String} The formatted string (i.e. `10630` becomes `7 days, 9 hours and 10 minutes`).
 */
const toString = (mins, opts) => {
    // The delta will reduce in scope as we move through the various time units.
    // Each time a time unit is matched, remove that unit from the delta.
    const parts = [];
    let delta = mins;

    // Determine the time period.

    // Are the minutes greater than a week?
    if (
        (delta >= WEEK || opts.display.inclusive) &&
        (opts.display.w || opts.display.ww)
    ) {
        parts.push(
            formatPart(Math.floor(delta / WEEK), 'w', opts.display.ww, opts)
        );
        delta -= Math.floor(delta / WEEK) * WEEK;
    }

    // Are the remaining(?) minutes greater than a day?
    if (
        (delta >= DAY || opts.display.inclusive) &&
        (opts.display.d || opts.display.dd)
    ) {
        parts.push(
            formatPart(Math.floor(delta / DAY), 'd', opts.display.dd, opts)
        );
        delta -= Math.floor(delta / DAY) * DAY;
    }

    // Are the remaining(?) minutes greater than an hour?
    if (
        (delta >= HOUR || opts.display.inclusive) &&
        (opts.display.h || opts.display.hh)
    ) {
        parts.push(
            formatPart(Math.floor(delta / HOUR), 'h', opts.display.hh, opts)
        );
        delta -= Math.floor(delta / HOUR) * HOUR;
    }

    // Are there any remaining minutes?
    if (
        (delta > 0 || opts.display.inclusive) &&
        (opts.display.m || opts.display.mm)
    ) {
        parts.push(formatPart(delta, 'm', opts.display.mm, opts));
    }

    if (parts.length === 1) {
        return parts.toString();
    }

    // Join parts with `delimiter`, other than the final one which should be `conjunction`.
    return `${parts.slice(0, parts.length - 1).join(opts.tokens.delimiter)}${
        opts.tokens.conjunction
    }${parts[parts.length - 1]}`;
};

/**
 * Create a new instance of minutes, scoped to an integer passed in.
 * @param  {Number} minutes The number of minutes this instance should work with.
 * @param  {Object} options The options to alter this particular instance of minutes.js
 * @return {Object}         The initialized instance of Minutes.
 */
export default (minutes, options = {}) => {
    // Attempt to parse the minutes as an integer.
    const mins = parseInt(minutes);

    // Only accept an integer.
    if ((Number.isNaN ?? isNaN)(mins)) {
        throw new Error('Minutes accepts an integer representing minutes.');
    }

    const opts = defaultOptions(options);

    return toString(mins, opts);
};
